packages/merge-conflicts/src/browser/merge-conflicts-provider.ts

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the 'License'); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { inject, injectable, postConstruct } from "inversify";
+import { MergeConflict } from "./merge-conflict";
+import { TextEditor, FileEditorTracker } from "@theia/editor/lib/browser";
+import { Emitter, Event } from "@theia/core";
+import { Deferred } from "@theia/core/lib/common/promise-util";
+import { MergeConflictsParser } from "./merge-conflicts-parser";
+
+@injectable()
+export class MergeConflictsProvider {
+
+    @inject(FileEditorTracker)
+    protected readonly tracker: FileEditorTracker;
+
+    @inject(MergeConflictsParser)
+    protected readonly mergeConflictParser: MergeConflictsParser;
+
+    protected readonly onDidUpdateEmitter = new Emitter<MergeConflicts>();
+    readonly onDidUpdate: Event<MergeConflicts> = this.onDidUpdateEmitter.event;
+
+    deferredValues = new Map<string, Deferred<MergeConflicts>>();
+    timeouts = new Map<string, number>();
+
+    @postConstruct()
+    protected initialize() {
```

Comment:
I would prefer to introduce more powerful abstraction like `ThrottlerDelayer` instead of `FileEditorTracker` and use it here:
```ts
    @postConstruct()
    protected initialize() {
        this.editorManager.onCreated(widget => {
            widget.editor.onDocumentContentChanged(() => 
                this.delayer.push(() => this.contentChanged(editor))
            );
            widget.disposed.connect(() => this.remove(editor));
        });
    }
```

This abstraction is more powerful and does not require changes in the editor extension.

Reply:
maybe we can use an existing library for it, if there is a clear winner, instead of implementing throttling and debouncing ourself

Reply:
Then I got you wrong. I'll revise it, I had it almost like this.

Reply:
funny, were chatting about this with @kittaakos. 
`lodash.throttle` seems to be a candidate.

Reply:
Cool, how many implementations do we have so far? :)

Reply:
We do a lot of debouncing, look for usages of `setTimeout` + `DisposableCollection`. About `throttling` not sure, probably it is the only place right now.

Reply:
scroll events in preview extension are throttled, too.